cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -324,17 +407,46 @@ Result<std::shared_ptr<io::InputStream>> GcsFileSystem::OpenInputStream(
     return Status::IOError("Only files can be opened as input streams");
   }
   ARROW_ASSIGN_OR_RAISE(auto p, GcsPath::FromString(info.path()));
-  return impl_->OpenInputStream(p);
+  return impl_->OpenInputStream(p.bucket, p.object, gcs::Generation(),
+                                gcs::ReadFromOffset());
 }
 
 Result<std::shared_ptr<io::RandomAccessFile>> GcsFileSystem::OpenInputFile(
     const std::string& path) {
-  return Status::NotImplemented("The GCS FileSystem is not fully implemented");
+  ARROW_ASSIGN_OR_RAISE(auto p, GcsPath::FromString(path));
+  auto metadata = impl_->GetObjectMetadata(p);
```

Comment:
I'm curious, does this add a roundtrip to the server?

Reply:
Yes, it does.  I am trying to ensure that `Read()` and `ReadAt()` and `Seek()` when going back are using the same generation of an object [*].  We could try to use undocumented (and likely to break) APIs to extract the generation without this roundtrip.  If it turns out the roundtrip (and I should add, the additional API charges) are really important, then I would rather add a documented API to the C++ client library and then use that here.


[*]: you probably know this, but objects in GCS are versioned.  You can have more than one version of the same object, and/or have the "latest" version replaced while you are reading from it.  I would think we want all operations in one `io::RandomAccessFile` to refer to the same generation.

Reply:
>  I would think we want all operations in one io::RandomAccessFile to refer to the same generation.

Oh, definitely. I just wonder if there's a way to extract the metadata from the first stream creation? Perhaps this is something that the GCS C++ library can provide (assuming the information exists at all at the HTTP level)?

Reply:
And thus my (maybe too obscure) note about "undocumented (and likely to break) APIs".  The C++ client library currently returns the HTTP headers, which include thinks like `x-goog-generation` (not the full metadata). This is only intended for debugging, and I **know** of future changes that will make these headers unavailable (sometimes, when the user selects a non-HTTP transport).

I can change the C++ client to return the generation in a future-proof API.  I would rather do that in a separate PR, after I fix the C++ client library (https://github.com/googleapis/google-cloud-cpp/issues/7677).

Reply:
> I can change the C++ client to return the generation in a future-proof API. I would rather do that in a separate PR, after I fix the C++ client library

No problem from me.